pkg/apis/certmanager/v1/const.go

```diff
@@ -28,7 +28,7 @@ const (
 	// minimum certificate duration before certificate expiration
 	MinimumRenewBefore = time.Minute * 5
 
-	// Default duration before certificate expiration if  Issuer.spec.renewBefore is not set
+	// Deprecated: the default is now 2/3 of Certificate's duration
 	DefaultRenewBefore = time.Hour * 24 * 30
```

Comment:
:+1:

Reply:
There's a lingering reference to this constant in 
```
git grep DefaultRenewBefore
pkg/apis/certmanager/v1/const.go:       DefaultRenewBefore = time.Hour * 24 * 30
pkg/apis/certmanager/v1alpha2/const.go: DefaultRenewBefore = time.Hour * 24 * 30
pkg/apis/certmanager/v1alpha3/const.go: DefaultRenewBefore = time.Hour * 24 * 30
pkg/apis/certmanager/v1beta1/const.go:  DefaultRenewBefore = time.Hour * 24 * 30
pkg/internal/apis/certmanager/fuzzer/fuzzer.go:                         s.Spec.RenewBefore = &metav1.Duration{Duration: v1.DefaultRenewBefore}

```

https://github.com/jetstack/cert-manager/blob/3242b83b12d3609ca8f9d6e540daef7e0c8803d7/pkg/internal/apis/certmanager/fuzzer/fuzzer.go#L29-L59

I think the purpose of that is to prevent the Duration and the RenewBefore from being "fuzzed" and so avoiding any invalid combinations.
So perhaps that code doesn't need changing until we eventually remove the constant.

Reply:
I guess I would be tempted to remove this part from fuzzer since the default in now dependent on what the duration of the issued cert is and is calculated internally.
I guess it could be set to 1/3 of the duration if that is set, but that too would not be correct since it is the actual duration that is used to calculate the renewBefore?

Reply:
Although we left the constant in, we're not actually using it, so this leaving it woudl be incorrect imo? What do you think?

Reply:
I removed defaulting for `renewBefore` from fuzzer in https://github.com/jetstack/cert-manager/pull/4092/commits/82dda9e469c6ad461d580e33e315c5c4e8bf85ce .
Happy to put it back if there are reasons to do that.

Reply:
I think my earlier comment about `avoiding any invalid combinations.` was wrong.
These fuzzed objects never get fed into the validation functions AFAIK.

They are only fed through the conversion functions, to ensure that many permutations of Certificates can be correctly converted forward and backward (roundtrip) between all available versions.

So next theory is that when there's a field which has been introduced in a later version, its value has to be fixed here, because it will sometimes be lost on backwards conversion and when converted forward again, the resulting Certificate will be different than the original.
With that said, I think that all of the currently supported API versions have both the Duration and RenewBefore fields, so it should be safe to remove the override values for both of them.
Try it and see if the tests still pass.

Reply:
@munnerz Please weigh in, if you have time.

Reply:
> With that said, I think that all of the currently supported API versions have both the Duration and RenewBefore fields, so it should be safe to remove the override values for both of them

The same about the other fields that are set in those fuzzer funcs such as `spec.DNSNames`. I see that this too is present in all API versions.

Reply:
`bazel test //...` passes even if I remove all of the defaulting from `pkg/internal/apis/certmanager/fuzzer/fuzzer.go`.

I guess it may make sense to remove the defaulting if it's no longer needed (probably in other fuzzer funcs too), but maybe not in this PR?

Reply:
Yeah, let's remove the rest (if appropriate) in another PR.

Reply:
#4093